File: src/main/java/xyz/oribuin/lilori/managers/TrackScheduler.java
package xyz.oribuin.lilori.managers;

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.LinkedBlockingQueue;

public class TrackScheduler extends AudioEventAdapter {
    private AudioPlayer player;
    private AudioPlayerManager playerManager;
    private final BlockingQueue<AudioTrack> queue;

    public TrackScheduler(AudioPlayer player, AudioPlayerManager playerManager) {
        this.player = player;
        this.playerManager = playerManager;
        this.queue = new LinkedBlockingQueue<>();

        if (player.getPlayingTrack() == null)
            nextTrack();
        ;
    }


    public void loadTrack(String song) {
        playerManager.loadItem(song, new AudioLoadResultHandler() {
            @Override
            public void trackLoaded(AudioTrack track) {
                System.out.println("Loaded! " + track.getInfo().title);
                player.startTrack(track, false);
            }

            @Override
            public void playlistLoaded(AudioPlaylist playlist) {
                nextTrack();
            }

            @Override
            public void noMatches() {
                nextTrack();
            }

            @Override
            public void loadFailed(FriendlyException exception) {
                exception.printStackTrace();
                nextTrack();
            }
        });
    }

    public void queue(AudioTrack track) {
        if (!player.startTrack(track, false))
            queue.offer(track);
    }


    public void nextTrack() {
        player.startTrack(queue.poll(), false);
    }

    @Override
    public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
        if (endReason.mayStartNext)
            nextTrack();
    }
}

File: src/main/kotlin/xyz/oribuin/lilori/manager/music/TrackScheduler.kt
package xyz.oribuin.lilori.manager.music

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import xyz.oribuin.lilori.LilOri
import xyz.oribuin.lilori.command.music.CmdLoop
import xyz.oribuin.lilori.handler.CommandHandler
import java.util.*

class TrackScheduler(private val bot: LilOri, private val player: AudioPlayer) : AudioEventAdapter() {
    private val queue: Queue<AudioTrack>
    fun queue(track: AudioTrack, override: Boolean) {
        if (!player.startTrack(track, !override)) {
            queue.offer(track)
        }
    }

    fun nextTrack() {
        player.startTrack(queue.poll(), true)
    }

    override fun onTrackEnd(player: AudioPlayer, audioTrack: AudioTrack, endReason: AudioTrackEndReason) {

        val command = bot.getManager(CommandHandler::class).getCommand("loop") as CmdLoop

        if (command.isLooping) {
            player.playTrack(audioTrack)
            return
        }

        if (endReason == AudioTrackEndReason.FINISHED)
            nextTrack()
    }

    init {
        player.volume = 75
        queue = LinkedList()
    }
}